### ISSUE TITLE
Keycloak Auth Provider too many redirects after login

### ISSUE THREAD

**Issue by UncleDinoso** [onesaitplatform/onesait-cloud-platform-community-deploy#18]
Cheers,

after deploying a fresh 6.0.2-CE instance with Keycloak, i get too many redirects after login. I configured Keycloak as your documentation explains. Unfortunately the documentation doesnt explain the client settings.

The user session is properly created and i can log in with the standard admin credentials.

![image](https://github.com/user-attachments/assets/cfe56091-af9a-4916-9638-50c401603387)

![image](https://github.com/user-attachments/assets/67c82d91-6fba-4e9c-8b80-4d5a5c20a092)

After login i have a circular redirect to "/controlpanel/login/oauth2/code/login"

In Keycloak logs, i see an Error: 2024-08-19 09:18:59,425 WARN  [org.key.events] (executor-thread-253) type=CODE_TO_TOKEN_ERROR, realmId=3c6957a4-1d37-4cef-a408-6c05f8711622, clientId=null, userId=null, ipAddress=172.20.0.1, error=client_not_found, grant_type=authorization_code

Can you give me a hint, what the problem is?

Thank you in advance!

Uncle Dinoso

**Comment by Fjgcornejo** [comment]
Hi,

I have tried to replicate what you stated on the issue unsuccessfully, as I was able to log in correctly.

However, I noticed that if I set "localhost" as the SERVER_NAME variable then I get the "Too many redirects" error, It seems that docker is not able to route the internal request to "localhost" in the final step for the authentication process, so this issue may be related to the SERVER_NAME variable.

Also, if you run "docker exec -it controlpanelservice sh" and then inside container "cd /var/log/platform-logs" there is a log for the controlpanel that may help to see what's happening. The warning log of Keycloak that you posted have nothing to do with this redirect problem.

Regards,
Javi.

**Comment by UncleDinoso** [comment]
Hy Javi, thank you for your reply. In the control-panel logs i didnt found any helpful logs.
I use the 21-ce Keycloak image from docker hub and in this article: https://onesaitplatform.atlassian.net/wiki/spaces/DOCT/pages/3579281411/Certificate+Authentication is described the main authentication flow, 'Multitenant-browser-flow', but i didnt have these flow. I try to build the flow but the conditions Vertical Choice Form and the others can be implemented, because the authentication provider is missing.
I also didnt have the opendId Scope, as you can see in the article https://onesaitplatform.atlassian.net/wiki/spaces/DOCT/pages/2220847130/Customize+token+Claims
Maybe the CE-image lacks configuration for using with the onesaitplatform?
At docker hub i can see the adding of onesaitplatform-keycloak-storage-provider
![image](https://github.com/user-attachments/assets/ca1a23b6-4378-4e98-b9b9-5ecd1832987b), but i didnt see authentication provider.

Is it correct, that on first startup of keycloak there is no realm for onesaitplatform exists and i have to configure all manually?

Regards,
Mario

**Comment by Fjgcornejo** [comment]
Hey!

Maybe you are missing the step where you need to run the op_module "keycloak-manager"?

After running Keycloak container itself, you should run the Keycloak-manager container. This module is in charge of synchronizing keycloak's database and our platform's database, this means that the realms and clients will be created with the authentication flows and provider that you mentioned.

So, for running this container, go to op_modules/keycloak-manager . You will need to modify .env file inside this folder first. Set you SERVER_NAME variable and most important, set the PLATFORM_ADMIN_APIKEY, which can be retrieved from the database (schema onesaitplatform_config).
<img width="973" alt="image" src="https://github.com/user-attachments/assets/c4fabc15-e5b6-4863-b889-8d5311ce68f9">

Then you can run "docker-compose up -d" for the module and look for the logs inside the container (/tmp directory), if everything went good, you should see something like this the first time:
<img width="1609" alt="image" src="https://github.com/user-attachments/assets/ecae7693-573d-48b3-8ea2-68054319c9db">

Let me know if it works or you have any issues in the process :)

Regards,
Javi.

**Comment by UncleDinoso** [comment]
Dear Javi,

thank you very much for the clarification. That is a missing point, i didn't understand, that i have to run the keycloak-manager module initially.

With your provided informations, i try to run the keycloak-manager, but it the configuration of keycloak doesnt start:
![image](https://github.com/user-attachments/assets/b44da2ae-42da-4a88-8653-2bbaa68c91ca)
The service will not be shutdown and run unfinite. (I restarted it one time)

I set the PLATFORM_ADMIN_APIKEY, the SERVER_NAME and KEYCLOAK_PASSWORD.
![image](https://github.com/user-attachments/assets/66653085-709c-4f21-9234-0e026078831e)

Maybe i have to truncate the keycloak database in configdb?

Regards,
Mario
